kyuubi-server/src/main/scala/org/apache/kyuubi/operation/BatchJobSubmission.scala

```diff
@@ -298,20 +298,22 @@ class BatchJobSubmission(
 
       MetricsSystem.tracing(_.decCount(MetricRegistry.name(OPERATION_OPEN, opType)))
 
-      if (!builder.processLaunched) {
+      // fast fail
+      if (isTerminalState(state)) {
+        killMessage = (false, s"batch $batchId is already terminal so can not kill it.")
         builder.close()
-        if (recoveryMetadata.isDefined) {
-          killMessage = killBatchApplication()
-        }
-        setState(OperationState.CANCELED)
-        updateBatchMetadata()
         return
       }
 
-      // fast fail
-      if (isTerminalState(state)) {
-        killMessage = (false, s"batch $batchId is already terminal so can not kill it.")
+      if (getBackgroundHandle == null) {
```

Comment:
we should trust the existed state machine to avoid use other flag to see the state. better to check with state `running` and `pending`

Reply:
But `BatchJobSubmission` is special, it will only change the state to `RUNNING` if get valid applicationInfo.

Reply:
shall we check `currentApplicationInfo` here and only kill if app id is not null ? I think it's no meaning to track the local state.

Reply:
seems we can check `if (state == OperationState.INITIALIZED) `

Reply:
have updated the code